src/cdp/connection.ts

```diff
@@ -0,0 +1,116 @@
+  private _browserClient: CdpClient;
+  private _clients: Map<string, CdpClient> = new Map();
+  private _commandCallbacks: Map<number, CdpCallbacks> = new Map();
+  private _nextId: number;
+
+  constructor(private _transport: IServer) {
+    this._nextId = 0;
+    this._transport.setOnMessage(this._onMessage);
+    this._browserClient = createClient(this, null);
+  }
+
+  /**
+   * Close the connection to the browser.
+   */
+  close() {
+    this._transport.close();
+    for (const [_id, { reject }] of this._commandCallbacks) {
+      reject(new Error('Disconnected'));
+    }
+    this._commandCallbacks.clear();
+    this._clients.clear();
+  }
+
+  /**
+   * @returns The CdpClient object attached to the root browser session.
+   */
+  browserClient(): CdpClient {
+    return this._browserClient;
+  }
+
+  /**
+   * Get a CdpClient instance by sessionId.
+   * @param sessionId The sessionId of the CdpClient to retrieve.
+   * @returns The CdpClient object attached to the given session, or null if the session is not attached.
+   */
+  sessionClient(sessionId: string): CdpClient | null {
+    return this._clients.get(sessionId) || null;
+  }
+
+  /**
+   * @internal
+   */
+  _sendCommand(
+    method: string,
+    params: {},
+    sessionId: string | null
+  ): Promise<{}> {
+    return new Promise((resolve, reject) => {
+      const id = this._nextId++;
+      this._commandCallbacks.set(id, { resolve, reject });
+      let messageObj: CdpMessage = { id, method, params };
+      if (sessionId) {
+        messageObj.sessionId = sessionId;
+      }
+
+      const messageStr = JSON.stringify(messageObj);
+      this._transport.sendMessage(messageStr);
+      logCdp('sent > ' + messageStr);
+    });
+  }
+
+  private _onMessage = async (message: string) => {
+    logCdp('received < ' + message);
+
+    const parsed = JSON.parse(message);
+
+    // Update client map if a session is attached or detached.
+    // Listen for these events on every session.
+    if (parsed.method === 'Target.attachedToTarget') {
+      const { sessionId } = parsed.params;
+      this._clients.set(sessionId, createClient(this, sessionId));
+    } else if (parsed.method === 'Target.detachedFromTarget') {
+      const { sessionId } = parsed.params;
+      const client = this._clients.get(sessionId);
+      if (client) {
+        this._clients.delete(sessionId);
+      }
+    }
+
+    if (parsed.id !== undefined) {
+      // Handle command response.
+      const callbacks = this._commandCallbacks.get(parsed.id);
+      if (callbacks) {
+        if (parsed.result) {
+          callbacks.resolve(parsed.result);
+        } else if (parsed.error) {
+          callbacks.reject(parsed.error);
+        }
+      }
+    } else if (parsed.method) {
+      const client = parsed.sessionId
+        ? this._clients.get(parsed.sessionId)
+        : this._browserClient;
+      if (client) {
+        client._onCdpEvent(parsed.method, parsed.params || {});
+      }
```

Comment:
Not all events have a `sessionId`. Events that don't have a `sessionId` belong to the browser session (the initial session you are attached to when you connect to the browser-wide websocket endpoint). It wouldn't be meaningful or useful to raise all events to the browser client. Say there's a CDP session X attached to page target and a `Page.frameAttached` event is received. That event and the frame IDs it contains are only meaningful in the context of session X.

The browser client is intended to send commands to the browser target only. So it can be used for things like accessing the `Browser` domain, creating new targets, and discovering/auto-attaching to top-level targets. It can't be used to communicate with these targets though, and it won't receive attachedToTarget events for nested targets like iframes and workers. 

To communicate with page targets, a client needs to know the corresponding `sessionId` and send commands to that specific session. That what's the session clients are for. it is also necessary to call `Target.setAutoAttach` once for every new CDP session so that sessions will be created for nested targets too.

In theory, we could have organized this a little differently. Just provide one big CdpClient object, where the sendCommand method takes an optional sessionId, and raised events have a sessionId too. However, this doesn't avoid the fact that CDP users need to be aware of the multiple CDP sessions and need to be explicit about which session they intend to communicate with. Having separate CdpClient objects for each session (including the browser session) is just a nice abstraction over this concept. Instead of the developer keeping track of sessionId strings, they keep track of CdpClients.

> The question above implies another one: we have the concept of CDP session, and the concept of BiDi session. Should they be the same? If yes, there will be a race condition after BiDi client is notified about creating new session and BiDi client attached to it.

These are not the same. A BiDi session is intended to represent an entire browser, and its entire tree of top-level pages, iframes, workers, etc. So it is an even higher-level abstraction on top of CDP, and will need to hide a lot of the complexity of CDP (e.g. session-per-target, separate targets for OOPIFs, etc.).

Reply:
I understand the concept of [`BiDi session`](https://w3c.github.io/webdriver-bidi/#command-session-subscribe) as a high-level abstraction correlated to the BiDi connection. 
> [A session is equivalent to a single instantiation of a particular user agent, including all its child browsers.](https://w3c.github.io/webdriver/#sessions)

From the CDP perspective, [`session` is created after attaching to `target`](https://chromedevtools.github.io/devtools-protocol/tot/Target/#method-attachToTarget). So it's an equivalent to a target (BiDi BrowsingContext) + specific client. 

What makes me confused is splitting clients by the CDP sessions. I don't quite understand how BiDi client's request should be routed to the certain `CdpClient`. There should be some kind of mapper `BrowsingContext`-`CdpSessionId`, and I don't understand where it will be.

Reply:
> There should be some kind of mapper BrowsingContext-CdpSessionId, and I don't understand where it will be.

That's correct. We will need some kind of mapper. When a new Context object is created (due to a Target.attachedToTarget event), we could get/create a CdpClient for the CDP sessionId and store this CdpClient as a property in the Context object. Then the Context object can route BiDi client's requests to this CdpClient.